ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixClusterManager.java

```diff
@@ -388,8 +387,86 @@ public List<PartitionId> getAllPartitionIds(String partitionClass) {
+          instanceName, e);
+      return null;
+    }
+
+    ReplicaId bootstrapReplica = isDataNodeInFullAuto ? getBootstrapReplicaInFullAuto(partitionIdStr, dataNodeId)
+        : getBootstrapReplicaInSemiAuto(partitionIdStr, dataNodeId);
+
+    // For now this method is only called by server which new replica will be added to. So if datanode equals to current
+    // node, we temporarily add this into a map (because we don't know whether store addition in storage manager
+    // succeeds or not). After store addition succeeds, current node is supposed to update InstanceConfig and will
+    // receive notification from Helix afterwards. At that time, dynamic cluster change handler will move replica from
+    // this map to clustermap related data structures that can be queried by other components.
+    if (bootstrapReplica != null && instanceName.equals(selfInstanceName)) {
+      // Note that this method might be called by several state transition threads concurrently.
+      bootstrapReplicas.put(partitionIdStr, bootstrapReplica);
+    }
+    return bootstrapReplica;
+  }
+
+  /**
+   * Gets bootstrap replica in Helix full auto mode. A disk with minimum usage will be used.
+   * @param partitionIdStr the partition id string
+   * @param dataNodeId the {@link DataNodeId} on which bootstrap replica is placed
+   * @return {@link ReplicaId} if there is a new replica satisfying given partition and data node. {@code null} otherwise.
+   */
+  private ReplicaId getBootstrapReplicaInFullAuto(String partitionIdStr, DataNodeId dataNodeId) {
+    String instanceName = getInstanceName(dataNodeId.getHostname(), dataNodeId.getPort());
+    // Check if data node or disk is in current cluster map, if not, set bootstrapReplica to null.
+    AmbryDataNode dataNode = instanceNameToAmbryDataNode.get(instanceName);
+    if (dataNode == null) {
+      logger.error(
+          "Datanode associated with bootstrap replica is not found in cluster map. Cannot create the replica.");
+      return null;
+    }
+
+    AmbryPartition mappedPartition =
+        new AmbryPartition(Long.parseLong(partitionIdStr), clusterMapConfig.clusterMapDefaultPartitionClass,
+            helixClusterManagerQueryHelper);
+    AmbryPartition currentPartition =
+        partitionNameToAmbryPartition.putIfAbsent(mappedPartition.toPathString(), mappedPartition);
+    if (currentPartition == null) {
+      logger.info("Partition {} is currently not present in cluster map, a new partition is created", partitionIdStr);
+      currentPartition = mappedPartition;
+    }
+
+    // Select a disk for the replica
+    Set<AmbryDisk> disks = ambryDataNodeToAmbryDisks.get(dataNode);
+    // Get disk which has maximum remaining space
+    AmbryDisk diskToUse = null;
+    for (AmbryDisk disk : disks) {
+      if (diskToUse == null || (disk.getRawCapacityInBytes() - disk.getUsedCapacityInBytes()
+          > diskToUse.getRawCapacityInBytes() - diskToUse.getUsedCapacityInBytes())) {
```

Comment:
Can we create a method call getRemainingCapacityInBytes()? or getFreeCapacityInBytes()?
And should we check the disk health here as well?

Reply:
Sure. Created method getAvailableSpaceInBytes(). 

Yeah, true. Like you said, we should check disk health too. Will add it in subsequent PR.